src/main/scala/edu/cmu/cs/obsidian/parser/Parser.scala

```diff
@@ -91,42 +91,47 @@ object Parser extends Parsers {
       }
     }
 
-    private def parseType: Parser[ObsidianType] = {
-
+    private def parseNonPrimitive: Parser[NonPrimitiveType] = {
+        opt(RemoteT()) ~ parseId ~ opt(LBracketT() ~ repsep(parseType, CommaT()) ~ RBracketT()) ~ opt(AtT() ~! parseIdAlternatives) ^^ {
+            case remote ~ id ~ genericParams ~ permissionToken => {
+                val isRemote = remote.isDefined
+                val actualGenerics = genericParams.map {
+                    case _ ~ params ~ _ => params
+                }.getOrElse(List())
 
-        val parseNonPrimitive: Parser[NonPrimitiveType] = {
-             opt(RemoteT()) ~ parseId ~ opt(AtT() ~! parseIdAlternatives) ^^ {
-                case remote ~ id ~ permissionToken => {
-                    val isRemote = remote.isDefined
-                    val typ = extractTypeFromPermission(permissionToken, id._1, isRemote, false)
-                    typ.setLoc(id)
-                }
+                val typ = extractTypeFromPermission(permissionToken, id._1, actualGenerics, isRemote, defaultOwned = false)
+                typ.setLoc(id)
             }
         }
+    }
 
+    private def parseType: Parser[ObsidianType] = {
         val intPrim = IntT() ^^ { t => IntType().setLoc(t) }
         val boolPrim = BoolT() ^^ { t => BoolType().setLoc(t) }
         val stringPrim = StringT() ^^ { t => StringType().setLoc(t) }
 
         parseNonPrimitive | intPrim | boolPrim | stringPrim
     }
 
-    private def extractTypeFromPermission(permission: Option[~[Token, Seq[Identifier]]], name: String, isRemote: Boolean, defaultOwned: Boolean): NonPrimitiveType = {
+    private def extractTypeFromPermission(permission: Option[~[Token, Seq[Identifier]]],
+                                          name: String, genericParams: Seq[ObsidianType],
+                                          isRemote: Boolean, defaultOwned: Boolean): NonPrimitiveType = {
         val defaultPermission = if (defaultOwned) Owned() else Inferred()
+
+        val contractType = ContractType(name, genericParams)
+
         permission match {
-            case None => ContractReferenceType(ContractType(name), defaultPermission, isRemote)
+            case None => ContractReferenceType(contractType, defaultPermission, isRemote)
             case Some(_ ~ permissionIdentSeq) =>
                 if (permissionIdentSeq.size == 1) {
                     val thePermissionOrState = permissionIdentSeq.head
-                    val permission = resolvePermission(thePermissionOrState._1)
```

Comment:
This refactor wasn't really necessary. In general, unnecessary changes make PRs harder to review. This change in particular takes work that was broken into two steps (for readability) and re-combines it into one, which is harder to read.